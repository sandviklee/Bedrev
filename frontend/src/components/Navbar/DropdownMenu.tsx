import * as DropdownMenu from "@radix-ui/react-dropdown-menu";
import styles from "./Dropdown.module.css";
import { Link, useLocation } from "react-router-dom";
import Hamburger from "hamburger-react";
import { useEffect, useState } from "react";

const Dropdown = () => {
  const [isOpen, setOpen] = useState<boolean>(false);
  const toggleOpen = () => {
    setOpen(!isOpen);
  };

  /**
   * Retrieves the auth token to
   * check if a user is logged in.
   */
  const [authToken, setAuthToken] = useState<string | null>(
    localStorage.getItem("auth-token"),
  );
  const location = useLocation();

  useEffect(() => {
    setAuthToken(localStorage.getItem("auth-token"));
  }, [location.state]);

  return (
    <DropdownMenu.Root open={isOpen} onOpenChange={toggleOpen}>
      <DropdownMenu.Trigger asChild>
        <button onClick={toggleOpen}>
          <Hamburger toggled={isOpen} />
        </button>
      </DropdownMenu.Trigger>
      <DropdownMenu.Portal>
        <DropdownMenu.Content
          className={styles.DropdownMenuContent}
          sideOffset={5}
        >
          <DropdownMenu.Item className="flex flex-row w-full justify-center py-1">
            <Link to="/">Hjem</Link>
          </DropdownMenu.Item>
          <DropdownMenu.Item className="flex flex-row w-full justify-center py-1">
            <Link to="/sok-bedrifter">SÃ¸k bedrifter</Link>
          </DropdownMenu.Item>
          {authToken == null ? (
            <DropdownMenu.Item className="flex flex-row w-full justify-center py-1">
              <Link to="/logg-inn">Logg inn</Link>
            </DropdownMenu.Item>
          ) : (
            <DropdownMenu.Item className="flex flex-row w-full justify-center py-1">
              <Link
                onClick={() => localStorage.removeItem("auth-token")}
                to="/"
              >
                Logg Ut
              </Link>
            </DropdownMenu.Item>
          )}

          <DropdownMenu.Separator className={styles.DropdownMenuSeparator} />
        </DropdownMenu.Content>
      </DropdownMenu.Portal>
    </DropdownMenu.Root>
  );
};

export default Dropdown;
