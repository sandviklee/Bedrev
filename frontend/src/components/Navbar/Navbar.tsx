import { Link, useNavigate, useLocation } from "react-router-dom";
import classNames from "classnames";
import Dropdown from "./DropdownMenu";
import bedrevLogo from "/bedrevLogo.svg";
import { useEffect, useState } from "react";
import { NavbarData } from "../../types/types";

/**
 * Dot component placed beside the
 * clicked navbar button (page).
 * @returns TSX component
 */
const Dot = () => {
  return (
    <>
      <div className="w-2 h-2 rounded-xl bg-purple"></div>
    </>
  );
};

interface NavbarProps {
  page?: "Hjem" | "Søk bedrifter" | "Logg Inn";
}

const navbarLinks = [
  {
    title: "Hjem",
    link: "/",
  },
  {
    title: "Søk bedrifter",
    link: "/sok-bedrifter",
  },
];

const buttonClass = classNames(
  "flex",
  "items-center",
  "justify-center",
  "h-[100%]",
  "w-[120px]",
  "font-whyte",
);

const Navbar = ({ page }: NavbarProps) => {
  const [authToken, setAuthToken] = useState<string | null>(
    localStorage.getItem("auth-token"),
  );
  const navigate = useNavigate();
  const location = useLocation();

  useEffect(() => {
    setAuthToken(localStorage.getItem("auth-token"));
  }, [location.state]);

  return (
    <nav className="flex flex-row w-full h-[120px] items-center px-[8%] text-base font-whyte">
      <Link to={`/`}>
        <img className="w-[150px]" src={bedrevLogo} alt="companylogo" />
      </Link>
      <div className="sm:visible sm:flex flex-row h-full w-full items-center justify-end hidden">
        {navbarLinks.map((data: NavbarData) => (
          <Link key={data.title} to={data.link} className={buttonClass}>
            <div
              className="pb-1.5 pr-2"
              style={{
                visibility: page == `${data.title}` ? "visible" : "hidden",
              }}
            >
              <Dot />
            </div>
            <p>{data.title}</p>
          </Link>
        ))}
        {authToken == null ? (
          <Link to={"/logg-inn"} className={buttonClass}>
            <div
              className="pb-1.5 pr-2"
              style={{
                visibility: page == `Logg Inn` ? "visible" : "hidden",
              }}
            >
              <Dot />
            </div>
            <p>Logg Inn</p>
          </Link>
        ) : (
          <Link
            to={"/"}
            onClick={async () => {
              await localStorage.removeItem("auth-token");
              navigate(-1);
            }}
            className={buttonClass}
          >
            <div
              className="pb-1.5 pr-2"
              style={{
                visibility: page == `Logg Inn` ? "visible" : "hidden",
              }}
            >
              <Dot />
            </div>
            <p>Logg ut</p>
          </Link>
        )}
      </div>
      <div className="sm:hidden flex w-full h-full items-center justify-end visible pb-4">
        <Dropdown />
      </div>
    </nav>
  );
};

export default Navbar;
