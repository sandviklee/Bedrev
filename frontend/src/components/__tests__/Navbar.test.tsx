import { describe, test, expect } from "vitest";
import Navbar from "../Navbar/Navbar.tsx";
import { render, screen } from "@testing-library/react";
import { BrowserRouter } from "react-router-dom";

describe("Navbar component", () => {
  test("renders navigation links correctly", () => {
    render(
      <BrowserRouter>
        <Navbar />
      </BrowserRouter>,
    );
    // The text is in a <p> tag, so we need to get the parent element which is the <a> tag
    const logoLink = screen.getByAltText("companylogo").closest("a");
    const homeLink = screen.getByText("Hjem").closest("a");
    const searchLink = screen.getByText("Søk bedrifter").closest("a");
    const loginLink = screen.getByText("Logg Inn").closest("a");

    expect(logoLink).toBeInTheDocument();
    expect(homeLink).toBeInTheDocument();
    expect(searchLink).toBeInTheDocument();
    expect(loginLink).toBeInTheDocument();
  });

  test("navigates to the correct routes", () => {
    render(
      <BrowserRouter>
        <Navbar />
      </BrowserRouter>,
    );
    // The text is in a <p> tag, so we need to get the parent element which is the <a> tag
    const logoLink = screen.getByAltText("companylogo").closest("a");
    const homeLink = screen.getByText("Hjem").closest("a");
    const searchLink = screen.getByText("Søk bedrifter").closest("a");
    const loginLink = screen.getByText("Logg Inn").closest("a");

    expect(logoLink?.getAttribute("href")).toBe("/");
    expect(homeLink?.getAttribute("href")).toBe("/");
    expect(searchLink?.getAttribute("href")).toBe("/sok-bedrifter");
    expect(loginLink?.getAttribute("href")).toBe("/logg-inn");
  });

  test("matches snapshot", () => {
    const { asFragment } = render(
      <BrowserRouter>
        <Navbar />
      </BrowserRouter>,
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
