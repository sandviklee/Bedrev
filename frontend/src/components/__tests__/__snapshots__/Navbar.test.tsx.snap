// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Navbar component > matches snapshot 1`] = `
<DocumentFragment>
  <nav
    class="flex flex-row w-full h-[120px] items-center px-[8%] text-base font-whyte"
  >
    <a
      href="/"
    >
      <img
        alt="companylogo"
        class="w-[150px]"
        src="/project2/bedrevLogo.svg"
      />
    </a>
    <div
      class="sm:visible sm:flex flex-row h-full w-full items-center justify-end hidden"
    >
      <a
        class="flex items-center justify-center h-[100%] w-[120px] font-whyte"
        href="/"
      >
        <div
          class="pb-1.5 pr-2"
          style="visibility: hidden;"
        >
          <div
            class="w-2 h-2 rounded-xl bg-purple"
          />
        </div>
        <p>
          Hjem
        </p>
      </a>
      <a
        class="flex items-center justify-center h-[100%] w-[120px] font-whyte"
        href="/sok-bedrifter"
      >
        <div
          class="pb-1.5 pr-2"
          style="visibility: hidden;"
        >
          <div
            class="w-2 h-2 rounded-xl bg-purple"
          />
        </div>
        <p>
          SÃ¸k bedrifter
        </p>
      </a>
      <a
        class="flex items-center justify-center h-[100%] w-[120px] font-whyte"
        href="/logg-inn"
      >
        <div
          class="pb-1.5 pr-2"
          style="visibility: hidden;"
        >
          <div
            class="w-2 h-2 rounded-xl bg-purple"
          />
        </div>
        <p>
          Logg Inn
        </p>
      </a>
    </div>
    <div
      class="sm:hidden flex w-full h-full items-center justify-end visible pb-4"
    >
      <button
        aria-expanded="false"
        aria-haspopup="menu"
        data-state="closed"
        id="radix-:r4:"
        type="button"
      >
        <div
          aria-expanded="false"
          class="hamburger-react"
          role="button"
          style="cursor: pointer; height: 48px; position: relative; transition: 0.4s cubic-bezier(0, 0, 0, 1); user-select: none; width: 48px; outline: none; transform: none;"
          tabindex="0"
        >
          <div
            style="background: currentColor; height: 3px; left: 8px; position: absolute; width: 32px; top: 13px; transition: 0.4s cubic-bezier(0, 0, 0, 1); transform: none;"
          />
          <div
            style="background: currentColor; height: 3px; left: 8px; position: absolute; width: 32px; top: 23px; transition: 0.4s cubic-bezier(0, 0, 0, 1); transform: none;"
          />
          <div
            style="background: currentColor; height: 3px; left: 8px; position: absolute; width: 32px; top: 33px; transition: 0.4s cubic-bezier(0, 0, 0, 1); transform: none;"
          />
        </div>
      </button>
    </div>
  </nav>
</DocumentFragment>
`;
